# Final Model Construction

Below is the construction for a multiple linear regression model predicting housing prices in King County. Data is imported from the preprocessed data that was cleaned in preceding notebooks and saved as separate csv files. The data is imported to this notebook and combined for the first model. Further iterations are contained in additional subsections of this notebook.

### Import Packages

In [1]:
import scipy.stats as stats
import pandas as pd
import statsmodels.api as sm
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.metrics import median_absolute_error, mean_squared_error 
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import LinearRegression


### Import and Combine Cleaned Data

In [2]:
categorical_ohe = pd.read_csv("./data/cat_hot_dataframe")
categorical_ordinal = pd.read_csv("./data/cat_ordinal_dataframe")
numeric = pd.read_csv("./data/initial_numeric_inputs")
target = pd.read_csv("./data/house_price_target_natlog")

### Construct Model

In [3]:
all_predictors = pd.concat([numeric, categorical_ohe], axis=1)
exog = sm.add_constant(all_predictors)
endog = target

first_model = sm.OLS(endog, exog).fit().summary()
first_model

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               ln_price   R-squared:                       0.648
Model:                            OLS   Adj. R-squared:                  0.648
Method:                 Least Squares   F-statistic:                     1655.
Date:                Tue, 25 Oct 2022   Prob (F-statistic):               0.00
Time:                        18:28:22   Log-Likelihood:                -5517.1
No. Observations:               21597   AIC:                         1.108e+04
Df Residuals:                   21572   BIC:                         1.128e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const                9.3377      0.090    103.931      0.000       9.162       9.514
bedrooms            -0.0374      0.003    -12.215      0.000      -0.043      -0.031
bathrooms            0.0871      0.005     17.387      0.000       0.077       0.097
ln_sqft_living       0.4146      0.014     29.751      0.000       0.387       0.442
sqft_lot          2.654e-07   7.41e-08      3.581      0.000     1.2e-07    4.11e-07
floors               0.1163      0.006     20.768      0.000       0.105       0.127
ln_sqft_above       -0.1776      0.012    -14.554      0.000      -0.202      -0.154
age                  0.0062    9.9e-05     62.310      0.000       0.006       0.006
renovated            0.0115      0.005      2.192      0.028       0.001       0.022
ln_sqft_living15     0.2520      0.011     23.643      0.000       0.231       0.273
sqft_lot15        -4.45e-07   1.13e-07     -3.926      0.000   -6.67e-07   -2.23e-07
x0_Fair             -0.1658      0.024     -6.792      0.000      -0.214      -0.118
x0_Good              0.0176      0.005      3.328      0.001       0.007       0.028
x0_Poor             -0.1696      0.059     -2.892      0.004      -0.284      -0.055
x0_Very Good         0.0856      0.009     10.005      0.000       0.069       0.102
x0_11 Excellent      0.2126      0.018     11.589      0.000       0.177       0.249
x0_12 Luxury         0.4726      0.035     13.600      0.000       0.404       0.541
x0_13 Mansion        0.7404      0.088      8.441      0.000       0.568       0.912
x0_3 Poor           -0.9715      0.313     -3.102      0.002      -1.585      -0.358
x0_4 Low            -1.1469      0.063    -18.342      0.000      -1.269      -1.024
x0_5 Fair           -1.0797      0.025    -42.922      0.000      -1.129      -1.030
x0_6 Low Average    -0.8892      0.016    -57.003      0.000      -0.920      -0.859
x0_7 Average        -0.6667      0.012    -53.750      0.000      -0.691      -0.642
x0_8 Good           -0.4518      0.011    -40.450      0.000      -0.474      -0.430
x0_9 Better         -0.2041      0.011    -18.049      0.000      -0.226      -0.182
==============================================================================
Omnibus:                       58.161   Durbin-Watson:                   1.964
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               75.080
Skew:                          -0.008   Prob(JB):                     4.97e-17
Kurtosis:                       3.288   Cond. No.                     7.44e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.44e+06. This might indicate that there are
strong multicollinearity or other numerical probl

### Evaluate Model Performance

 (JB):score is fairly low, some of our data skew may be impacting our results
 also Cond. No.	7.44e+06 is very large, indicating significant multicollinearity  
 our Rsqr at .648 is decent but will continue to improve it.


### Assess Potential Model Improvements

(1)We're going to scale our features, so they have equal impact on their outcome
(2)Assess feature for collinearity and remove as needed


# 2nd Model

In [4]:
ss = StandardScaler()
ss.fit(numeric)
num_scaled = ss.transform(numeric)